File: 프로그래머스/3/43238. 입국심사/입국심사.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

typedef long long ll;

bool check(vector<int> times, int n, ll v) {
    ll cnt = 0;
    for(int i=0; i<times.size(); i++) {
        cnt += (v / times[i]);
    }
    return cnt >= n;
}

long long solution(int n, vector<int> times) {
    ll st = 1;
    sort(times.begin(), times.end());
    ll end = (n) * (ll) times.back();
    
    while(st < end) {
        ll mid = (st+end)/2;
        if(check(times, n, mid)) end = mid;
        else st = mid + 1;
    }
    
    return st;
}

File: 프로그래머스/3/43238. 입국심사/입국심사.java
class Solution {
    public long solution(int n, int[] times) {
        long left = 1;
        long right = 1;
        
        for(int time : times) {
            right = Math.max(right, time);
        }
        
        right = right * n;
        long answer = right;
        
        while(left <= right) {
            long mid = (left + right) / 2;
            long count = 0;
            
            for(int time : times) {
                count += (mid / time);
            }
            
            if(count >= n) {
                answer = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return answer;
    }
}